src/GitServer/GitServer.ts

```diff
@@ -0,0 +1,298 @@
+import Path from 'path'
+import http from 'http'
+import { promisify } from "util";
+import { AddressInfo } from "net";
+import { Readable } from "stream";
+import { parse } from 'querystring'
+import { HttpDuplex } from "./HttpDuplex";
+import uploadPack from "./upload-pack/UploadPack";
+import VaultStore from "@polykey/VaultStore/VaultStore";
+
+// Here is the protocol git outlines for sending pack files over http:
+// https://git-scm.com/docs/pack-protocol/2.17.0
+// https://github.com/git/git/blob/master/Documentation/technical/pack-protocol.txt
+// This should be consulted in developing our upload pack implementation
+
+// This git backend (as well as HttpDuplex class) is heavily inspired by node-git-server:
+// https://github.com/gabrielcsapo/node-git-server
+
+const services = ['upload-pack', 'receive-pack']
+
+class GitServer {
+  repoDir: string;
+  vaultStore: VaultStore;
+  autoCreate = false
+  server: http.Server;
+  constructor(
+    repoDir: string,
+    vaultStore: VaultStore
+  ) {
+    this.repoDir = repoDir
+    this.vaultStore = vaultStore
+  }
+
+  /**
+   * Find out whether `repoName` exists.
+   */
+  exists(repo: string) {
+    // TODO: consider if vault has been shared
+    return fs.existsSync(Path.join(this.repoDir, repo))
+  }
+
+  /**
+   * returns the typeof service being process
+   */
+  getType(service: any): string {
+    switch (service) {
+      case 'upload-pack':
+        return 'fetch'
+      case 'receive-pack':
+        return 'push'
+      default:
+        return 'unknown'
+    }
+  }
+
+
+  // alternatively you can also pass authenticate a promise
+  authenticate(
+    type: string,
+    repo: string,
+    username: string,
+    password: string,
+    headers: http.IncomingHttpHeaders
+  ): boolean {
+    if(username === 'foo') {
+      return true
+    }
+    return false
+  }
+
+  /**
+   * Handle incoming HTTP requests with a connect-style middleware
+   */
+  handle(req: http.IncomingMessage, res: http.ServerResponse) {
+
+    const infoHandler = (req: http.IncomingMessage, res: http.ServerResponse): boolean => {
+
+      if (req.method !== 'GET') return false
+
+      const u = url.parse(req.url!)
+      const m = u.pathname!.match(/\/(.+)\/info\/refs$/)
+      if (!m) return false
+      if (/\.\./.test(m[1])) return false
+
+
+      const repo = m[1]
+      const params = parse(u.query!)
+      if (!params.service) {
+        res.statusCode = 400
+        res.end('service parameter required')
+        return true
+      }
+
+      const service = (<string>params.service).replace(/^git-/, '')
+      if (services.indexOf(service) < 0) {
+        res.statusCode = 405
+        res.end('service not available')
+        return true
+      }
+
+      const repoName = this.parseGitName(m[1])
+
+      // check if the repo is authenticated
+      const type = this.getType(service)
+      const headers = req.headers
+      const { username, password } = {username: 'foo', password: 's'}
```

Comment:
These are definitely stubs right?

Reply:
yeah these were just stubs, I am yet to dive deep into mutual TLS authentication. I will begin looking into PKIjs for a self signed certificate. I think its good if we can add it as a part of this PR. Then later if we want to give users the ability to sign with their own internal CA, then we can do that.

Reply:
As discussed, PKI can come later but for now I was looking into having a TLS connection with a self-signed certificate and I have created a new `ConnectionManager` class for this. The idea being we have one connection and multiplex the git http stuff over that. I have had some trouble trying to get http working over an existing socket (i.e. a tls socket) but I haven't found an existing way to do it. I suspect we probably want to build our own HTTP parser or use an existing one. I will create an issue for this.

Reply:
https://github.com/MatrixAI/js-polykey/issues/46